consensus/src/consensus/remote_event_dispatcher.rs

```diff
@@ -0,0 +1,315 @@
+                if !state.read().contains_peer(&peer_id) {
+                    return ResponseSubscribeToAddress {
+                        result: Err(InvalidOperation),
+                    };
+                }
+
+                state
+                    .write()
+                    .remove_addresses(&peer_id, self.addresses.clone());
+            }
+        }
+        ResponseSubscribeToAddress { result: Ok(()) }
+    }
+}
+
+/// The state that is mantained by the remote event dispatcher:
+/// essentially the addresses and peers that are subscribed to us.
+pub struct RemoteEventDispatcherState<N: Network> {
+    /// HashMap containing a mapping from peers to their interesting addresses
+    pub subscribed_peers: HashMap<N::PeerId, HashSet<Address>>,
+
+    /// Mantains the current list of interesting addresses and the peers that are interested in those addresses
+    pub subscriptions: HashMap<Address, HashSet<N::PeerId>>,
+}
+
+impl<N: Network> RemoteEventDispatcherState<N> {
+    /// Creates a new state, only state should be needed per client.
+    pub fn new() -> Self {
+        Self {
+            subscribed_peers: HashMap::new(),
+            subscriptions: HashMap::new(),
+        }
+    }
+
+    /// Returns if a peer is part of our state.
+    pub fn contains_peer(&self, peer_id: &N::PeerId) -> bool {
+        self.subscribed_peers.contains_key(peer_id)
+    }
+
+    /// Returns the number of peers that are currently subscribed to us.
+    pub fn number_of_peers(&self) -> usize {
+        self.subscribed_peers.len()
+    }
+
+    /// Adds a new address for an specific peer.
+    pub fn add_addresses(&mut self, peer_id: &N::PeerId, addresses: Vec<Address>) {
+        // If we already knew this peer, then we just update its interesting addresses
+        if let Some(interesting_addresses) = self.subscribed_peers.get_mut(peer_id) {
+            interesting_addresses.extend(addresses.iter().cloned())
+        } else {
+            // Otherwise, we insert a new entry for this peer
+            self.subscribed_peers
+                .insert(*peer_id, HashSet::from_iter(addresses.iter().cloned()));
+        }
+
+        // Now we update our address mapping
+        for address in addresses {
+            if let Some(peers) = self.subscriptions.get_mut(&address) {
+                peers.insert(*peer_id);
+            } else {
+                let mut new_peers = HashSet::new();
+                new_peers.insert(*peer_id);
+                self.subscriptions.insert(address, new_peers);
+            }
+        }
+    }
+
+    /// Removes a peer from the state, used when a peer leaves the network.
+    pub fn remove_peer(&mut self, peer_id: &N::PeerId) {
+        if let Some(peer_addresses) = self.subscribed_peers.get(peer_id) {
+            // Obtain the addresses that are interested to this peer and remove the peer from those addresses.
+            peer_addresses.into_iter().for_each(|address| {
```

Comment:
Clippy warning here: this `.into_iter()` call is equivalent to `.iter()` and will not consume the `HashSet`